File: crates/guide-core/src/models/document.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::shared::IngestionStatus;

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema, Default)]
#[serde(rename_all = "snake_case")]
pub enum DocumentKind {
    DmGuide,
    MonsterManual,
    Srd,
    #[default]
    Campaign,
    Supplemental,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct CampaignDocument {
    pub id: Uuid,
    pub campaign_id: Uuid,
    pub filename: String,
    pub file_size_bytes: i64,
    pub stored_path: String,
    pub page_count: Option<i32>,
    pub document_kind: DocumentKind,
    pub description: Option<String>,
    pub ingestion_status: IngestionStatus,
    pub ingestion_error: Option<String>,
    pub story_extraction_status: String,
    pub story_extraction_error: Option<String>,
    pub ingestion_progress: Option<String>,
    pub uploaded_at: DateTime<Utc>,
    pub ingested_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct GlobalDocument {
    pub id: Uuid,
    pub title: String,
    pub filename: String,
    pub file_size_bytes: i64,
    pub stored_path: String,
    pub page_count: Option<i32>,
    pub document_kind: DocumentKind,
    pub ingestion_status: IngestionStatus,
    pub ingestion_error: Option<String>,
    pub ingestion_progress: Option<String>,
    pub uploaded_at: DateTime<Utc>,
    pub ingested_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct RankedChunk {
    pub content: String,
    pub section_path: String,
    pub doc_title: String,
    pub score: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct DocSummary {
    pub doc_id: Uuid,
    pub doc_name: String,
    pub filename: String,
    pub summary: String,
    pub scope: String,
    pub ingested_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema, Default)]
pub struct MetaIndex {
    pub scope: String,
    pub entries: Vec<DocSummary>,
}

/// A single entry from a document's Table of Contents.
#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct TocEntry {
    pub title: String,
    /// 1-based page number as printed in the document.
    pub page: u32,
    /// Nesting depth: 0 = top-level chapter, 1 = section, 2 = subsection, …
    pub depth: u8,
}

/// A single entry from a document's back-of-book Index.
#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct IndexEntry {
    pub term: String,
    /// Page numbers where this term appears.
    pub pages: Vec<u32>,
}

File: src/guide/models/document.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel

from .shared import IngestionStatus


class DocumentKind(str, Enum):
    campaign = "campaign"
    rulebook = "rulebook"


class CampaignDocument(BaseModel):
    id: UUID
    campaign_id: UUID
    filename: str
    file_size_bytes: int = 0
    stored_path: str
    page_count: int | None = None
    document_kind: DocumentKind = DocumentKind.campaign
    ingestion_status: IngestionStatus = IngestionStatus.pending
    ingestion_error: str | None = None
    uploaded_at: datetime
    ingested_at: datetime | None = None


class GlobalDocument(BaseModel):
    id: UUID
    title: str
    filename: str
    file_size_bytes: int = 0
    stored_path: str
    page_count: int | None = None
    ingestion_status: IngestionStatus = IngestionStatus.pending
    ingestion_error: str | None = None
    uploaded_at: datetime
    ingested_at: datetime | None = None


class RankedChunk(BaseModel):
    content: str
    section_path: str = ""
    doc_title: str = ""
    score: float = 0.0


class DocSummary(BaseModel):
    doc_id: UUID
    doc_name: str
    filename: str
    summary: str  # empty string if LLM unavailable
    scope: str    # campaign_id string OR "global"
    ingested_at: datetime


class MetaIndex(BaseModel):
    scope: str
    entries: list[DocSummary] = []
